doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,412 @@
+    solution proposal below).
+
+-   Dealing with Webhook sources. These are fundamentally different than the other
+    storage-owned source types whose operators implement the `SourceRender` trait,
+    and they do not contain a 'progress' collection. While they may also be
+    converted to use 'tables' in the future, that work is not in-scope for this proposal.
+
+## Solution Proposal
+
+The concept of a _source_ will be unified around a single relation representing
+the _progress_ of that source's ingestion pipeline. A query of any _source_
+object (of any type) will return the data that is currently in the
+`<source>_progress` collection.
+
+We have been investing in the user experience around _tables_, such that
+replacing _subsources_ with _tables_ would reduce the overall
+mental overhead for new users.
+
+By allowing tables to be marked as _read-only_, we can easily migrate
+existing subsources to be tables without needing to figure out how to merge
+`INSERT/UPDATE/DELETE` statements with the data written by their corresponding source.
+
+The default `CREATE SOURCE ..` statement will just create a top-level source object
+that represents a single ingestion of data from an upstream system. To actually
+ingest data from that upstream system into a persist collection, the user will
+use a `CREATE TABLE .. FROM SOURCE` statement with an option that references the source
+and the external reference.
+
+We will allow more than one table to reference the same external upstream reference,
+using a potentially different set of columns. This allows a user to handle an upstream
+schema change by creating a new table for the same upstream table and then performing
+a blue-green swap operation to switch their downstream dependencies to the new table,
+and then drop the old one. While we could also enable multiple subsources to
+reference the same external upstream reference, we will instead continue to restrict this
+to provide an incentive for users to opt-in to the migration to tables and allow us
+to deprecate subsources sooner.
+
+The existing options to `CREATE SOURCE` for Postgres & MySQL sources that specify
+automatic subsources to create (`FOR ALL TABLES`, `FOR SCHEMAS` and `FOR TABLES`) will be
+removed. Instead, all upstream tables to ingest from this source will need to be
+explicitly referenced in a `CREATE TABLE` statement. While this may seem controversial,
+more often than not these options cause upstream data that does not need to be
+brought into Materialize to be ingested, and by using explicit statements for each
+table to be ingested it makes Materialize configuration much more amenable to
+object<->state mappings in tools like dbt and Terraform.
+
+We will instead introduce a SQL statement to return all known upstream tables
+similar for a given source, optionally filtered by schema(s). This will allow
+the web console to expose a guided 'create source' workflow for users where
```

Comment:
We should we include the barebones `CREATE TABLE...FROM SOURCE` DDL statement in the output of this command, if possible.